gunrock/app/mf/mf_test.cuh

```diff
@@ -7,324 +7,605 @@
 
-    return elapsed;
 #endif
+    
+    return elapsed;
 }
 
 /**
- * @brief Validation of SSSP results
- * @tparam     GraphT        Type of the graph
- * @tparam     ValueT        Type of the distances
- * @param[in]  parameters    Excution parameters
- * @param[in]  graph         Input graph
- * @param[in]  src           The source vertex
- * @param[in]  h_distances   Computed distances from the source to each vertex
- * @param[in]  h_preds       Computed predecessors for each vertex
- * @param[in]  ref_distances Reference distances from the source to each vertex
- * @param[in]  ref_preds     Reference predecessors for each vertex
- * @param[in]  verbose       Whether to output detail comparsions
- * \return     GraphT::SizeT Number of errors
+ * @brief Validation of MF results
+ *
+ * @tparam     GraphT	      Type of the graph
+ * @tparam     ValueT	      Type of the distances
+ *
+ * @param[in]  parameters     Excution parameters
+ * @param[in]  graph	      Input graph
+ * @param[in]  source	      The source vertex
+ * @param[in]  sink           The sink vertex
+ * @param[in]  h_flow	      Computed flow on edges 
+ * @param[in]  ref_flow	      Reference flow on edges
+ * @param[in]  verbose	      Whether to output detail comparsions
+ *
+ * \return     int  Number of errors
  */
-template <
-    typename GraphT,
-    typename ValueT = typename GraphT::ValueT>
-typename GraphT::SizeT Validate_Results(
-             util::Parameters &parameters,
-             GraphT           &graph,
-    // TODO: add problem specific data for validation, e.g.:
-    	typename GraphT::VertexT   src,
-                    ValueT   *h_excess,
-         ValueT   *ref_excess = NULL,
-	 bool      verbose       = true)
+template <typename GraphT, typename ValueT, typename VertexT>
+int Validate_Results(
+	util::Parameters  &parameters,
+        GraphT		  &graph,
+    	VertexT		  source,
+    	VertexT		  sink,
+        ValueT		  *h_flow,
+	VertexT		  *reverse,
+	int		  *min_cut,
+        ValueT		  *ref_flow = NULL,
+	bool		  verbose = true)
 {
-    typedef typename GraphT::VertexT VertexT;
-    typedef typename GraphT::SizeT   SizeT;
-    // TODO: change to other representation, if not using CSR
-    typedef typename GraphT::CsrT    CsrT;
+    typedef typename GraphT::CsrT   CsrT;
+    typedef typename GraphT::SizeT  SizeT;  
 
-    SizeT num_errors = 0;
+    int num_errors = 0;
     bool quiet = parameters.Get<bool>("quiet");
+    auto nodes = graph.nodes;
+    auto edges = graph.edges;
+
+    auto e_start = graph.CsrT::GetNeighborListOffset(sink);
+    auto num_neighbors = graph.CsrT::GetNeighborListLength(sink);
+    auto e_end = e_start + num_neighbors;
+    ValueT flow_incoming_sink = 0;
+    for (auto e = e_start; e < e_end; ++e)
+    {
+	auto flow_e_in = h_flow[reverse[e]];
+	if (util::isValid(flow_e_in))
+	    flow_incoming_sink += flow_e_in;
+    }
+    util::PrintMsg("Max Flow GPU = " + std::to_string(flow_incoming_sink));
+
+    // Verify min cut h_flow
```

Comment:
I think we need another look at Validation, on larger dataset like `hollywood-2009` or smaller ones like `cheaseapke`:

```
Run 0, elapsed: 114.709139 ms, #iterations = 5
Max Flow GPU = 0.000000
MIN CUT flow = 4.000000
FAIL: Min cut 4.000000 and max flow 0.000000 are not equal
Flow Validity (compare results):
2 errors occurred.
```

and

```
-----------------------------------
Run 0, elapsed: 1.435995 ms, #iterations = 11
Max Flow GPU = 11.000000
MIN CUT flow = 1.000000
FAIL: Min cut 1.000000 and max flow 11.000000 are not equal
Flow Validity (compare results):
1 errors occurred.
```

Are you comparing the right flows? Why is min cut being compared to max flow, from my understanding the two have different outputs.

Reply:
Please let me know what the command-line produces that.

Reply:
1st run: `./bin/test_mf_10.0_x86_64 market /data-2/gunrock-dataset/large/hollywood.mtx --source=0 --sink=100``

2nd run: `./bin/test_mf_10.0_x86_64 market ../../dataset/small/chesapeake.mtx --source=0 --sink=39`